### PR TITLE
Enable optimizer in backward in TorchTitan

### PR DESCRIPTION
**Pull request by mori360**
Add `--training.enable_optimizer_in_backward` to enable optimizer in the backward
With such feature, `optim.step()` and `optim.zero_grad()` can be conducted during `loss.backward()`, optimizing the peak memory at training.

Caution: optimizer in the backward would free the gradients during backward, could not work with gradient clipping.

Results:
TL;DR: optimizer_in_backward could achieve a memory saving around 3GiB in llama3 test .toml with slightly improvements on tps and mfu
We conduct tests based on `llama3_8b.toml` by running `CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh`.
During the tests, we change `compile`, `enable_optimizer_in_backward`, `activation_checkpoint.mode` to show the performance raised by optimizer_in_backward in loggings on memory cost, tps and mfu.
Screenshot 2024-12-02 at 3 59 41 PM

Analysis:
When using optim in backward, the amount of memory reserved to the reduce-scatter stream is smaller, causing a lower **reserved memory** shown in the table, around 2-3 GiB
Take `activation_checkpoint.mode = selective` as an example: 
Screenshot 2024-12-02 at 3 31 46 PM

Also, when `activation_checkpoint.mode = full`, there's around `(num_params-2)*(param_grad_mem) - peak_memory_at_forward` decrease in **active memory**, around 1.2GiB in the figure below.
Screenshot 2024-12-02 at 2 57 01 PM

### CODE REVIEW

File: torchtitan/checkpoint.py [hunk 1]
```diff
@@ -103,24 +103,41 @@ def __init__(
         self,
         model: Union[nn.Module, List[nn.Module]],
         optim: Union[torch.optim.Optimizer, List[torch.optim.Optimizer]],
```
Comment (fegin) [inline torchtitan/checkpoint.py]: What would be the typing for `optim` after supporting `optim_in_bwd`?

Comment (fegin) [reply torchtitan/checkpoint.py]: That would help me to understand the code.

Comment (mori360) [reply torchtitan/checkpoint.py]: Thank you for the comments. `optim` after `optim_in_bwd` would be a `List[Optimizer]`
For the general `optim`, it's based from `torch.optim.AdamW(model.parameters())`
While after `optim_in_bwd`, each `param` in `model.parameters()` would generate an sub optimizer `torch.optim.AdamW(param)`, there's a list with length of `len(model.parameters())` of sub optimizers
The sub optimzier here could call the hook separately to minimize the peak memory in gradient calculation.(to free them earlier)

Normally, the `optim` after `optim_in_bwd` would be a structure like `Dict{param: sub_optim(param)}`, here to simplify the code, just leave as `List[sub_optim]`

File: torchtitan/config_manager.py [hunk 2]
```diff
@@ -270,6 +270,14 @@ def __init__(self):
             action="store_true",
             help="Whether to apply loss parallel when sequence parallel is enabled",
         )
+        self.parser.add_argument(
+            "--training.enable_optimizer_in_backward",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]: maybe put in `optimizer` instead of `training`

File: torchtitan/config_manager.py [hunk 3]
```diff
@@ -270,6 +270,14 @@ def __init__(self):
...
+            type=bool,
+            default=False,
+            help="""
+            Whether to apply optimizer in the backward. Caution, optimizer_in_backward
+            cannot compile with gradients clipping.""",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]:
```suggestion
            is not compatible with gradients clipping.""",
```

**Review by tianyu-l — COMMENTED** [review]
discussed offline: we should use inheritance or other design patterns to separate the code and make it more readable.

File: torchtitan/checkpoint.py [hunk 4]
```diff
@@ -123,6 +123,30 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
         list(map(func, self.model, self.optim))
 
 
+class OptimizerInBackwardWrapper(OptimizerWrapper):
```
Comment (fegin) [inline torchtitan/checkpoint.py]: I'm not sure if we want another wrapper. Can't we just flatten the `OptimizerContainer` in `OptimizerWrapper.__init__`?

File: torchtitan/optimizer.py [hunk 5]
```diff
@@ -40,6 +41,45 @@ def _build_optimizer(model):
...
+        else:
+            raise NotImplementedError(f"Optimizer {name} not added.")
+
+        def optim_hook(param) -> None:
+            optim_dict[param].step()
+            optim_dict[param].zero_grad()
+
+        for param in model.parameters():
+            if param.requires_grad:
+                param.register_post_accumulate_grad_hook(optim_hook)
```
Comment (fegin) [inline torchtitan/optimizer.py]: We probably need to document that users could not call `register_post_accumulate_grad_hook` after the optimizer is built. Otherwise, the gradients are clear and users' `register_post_accumulate_grad_hook` will be useless.

Comment (mori360) [reply torchtitan/optimizer.py]: Added caution in config_manager

File: torchtitan/checkpoint.py (hunk 4 above)
Comment (mori360) [reply torchtitan/checkpoint.py]: Flatten optims here so could follow the same get/set state_dict logic

File: torchtitan/checkpoint.py [hunk 6]
```diff
@@ -102,10 +103,23 @@ class OptimizerWrapper(Stateful):
...
+                for sub_optim in optim_group
+            ]
+        else:
+            optimizers = optim.optimizers
+            self.optim = (
+                [optimizers]
+                if isinstance(optimizers, torch.optim.Optimizer)
+                else optimizers
+            )
+        print(self.optim)
```
Comment (fegin) [inline torchtitan/checkpoint.py]: Remove this print or change to logger.info if we need to add some information.

**Review by fegin — APPROVED** [review]
We can land this PR first but we should consider have only one OptimizerWrapper and remove the one in checkpoint.

File: torchtitan/checkpoint.py [hunk 7]
```diff
@@ -165,7 +178,7 @@ def __init__(
         self,
         dataloader: DataLoader,
         model_parts: List[nn.Module],
-        optimizers: List[torch.optim.Optimizer],
+        optimizers: OptimizersContainer,
         lr_schedulers: List[torch.optim.lr_scheduler.LRScheduler],
```
Comment (tianyu-l) [inline torchtitan/checkpoint.py]: wouldn't the checkpointing wrapping strategies also be different for the two cases? E.g. for optimizer in bwd, the `lr_schedulers` here seems to be a **List of List**?

File: torchtitan/optimizer.py [hunk 8]
```diff
@@ -79,27 +118,44 @@ def linear_warmup_linear_decay(
...
+            schedulers.step()
+
+
+class SchedulersInBackwardContainer:
+    """Util for calling step on multiple learning rate schedulers when optimizers are in backward"""
+
+    def __init__(self, optimizers, lr_lambda):
+        self.schedulers = []
+        for optimizer in optimizers:
+            schedulers = []
```
Comment (tianyu-l) [inline torchtitan/optimizer.py]: naming is a bit confusing, let's differentiate schedulers names for `model_parts` and for params in a `model_part`

File: torchtitan/config_manager.py [hunk 9]
```diff
@@ -194,6 +194,15 @@ def __init__(self):
             action="store_true",
             help="Whether the fused implementation(CUDA only) is used.",
         )
+        self.parser.add_argument(
+            "--optimizer.backward",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]: I somehow think we can have a better name for this. `backward` is less informative.
How about `early_step_in_backward` or `async_step`. cc: @fegin

File: torchtitan/config_manager.py [hunk 10]
```diff
@@ -194,6 +194,15 @@ def __init__(self):
...
+            type=bool,
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]: we can use `action="store_true",` since it's a boolean flag. Please check other configs as examples.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/checkpoint.py (hunk 7 above)
Comment (mori360) [reply torchtitan/checkpoint.py]: changed similar to optimizers as `SchedulersContainer`
with the modification inside `SchedulersContainer`, `SchedulersInBackwardContainer.schedulers` would only output the first scheduler in each list so that it would return a **List**, same as `SchedulersContainer.schedulers`

File: torchtitan/config_manager.py (hunk 9 above)
Comment (mori360) [reply torchtitan/config_manager.py]: Use `early_step_in_backward` right now

File: torchtitan/optimizer.py (hunk 8 above)
Comment (mori360) [reply torchtitan/optimizer.py]: Currently use optimizers -> optim_group -> sub_optim
We could synchronize the names in the next PR when we clean up the optimizer wrappers

File: torchtitan/optimizer.py [hunk 11]
```diff
@@ -79,27 +118,50 @@ def linear_warmup_linear_decay(
...
+
+    def __init__(self, optimizers, lr_lambda):
+        # all the schedulers for each optimizer group are the same, here we only store the first one
+        # to self.schedulers follow the same structure as SchedulersContainer, but store all of them
+        # to self.all_schedulers for container.step() to call
+        self.schedulers = []
+        self.all_schedulers = []
+        for optim_group in optimizers:
+            if len(optim_group) > 0:
+                self.schedulers.append(LambdaLR(optim_group[0], lr_lambda=lr_lambda))
```
Comment (tianyu-l) [inline torchtitan/optimizer.py]: this doesn't sound right -- It seems what you store in the `state_dict` is not the one used in training. E.g. after several steps, the scheduler in training would decrease the learning rate, but the one you created here would still have original LR.

I still think we should address this by unifying "container" and "wrapper". This way the `state_dict` of `SchedulersInBackwardContainer` can just save the `self.all_schedulers[0][0]` or something.

I'm OK with having a separate PR for it, but here you'll need to change `self.schedulers` to use the objects in `self.all_schedulers` rather than creating its own.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/optimizer.py (hunk 11 above)
Comment (mori360) [reply torchtitan/optimizer.py]: Yeah, the logic here is not right, temporally using `self.all_schedulers[0][0]` like expression under isinstance(lr_schedulers, SchedulersInBackwardContainer).
Would have a separate PR to restructure optimzierwrapper and schedulerwrapper.

**Review by tianyu-l — APPROVED** [review]
thanks for the work! let's address the followups in another PR
